### PR TITLE
Fix #566: serialize commit to hex

### LINKED ISSUE #566: commitment serialization format in utxo api endpoint
We should serialize the commitment as a hex string for convenience (currently serialized as a byte array in the json itself) - 

```
curl http://127.0.0.1:13413/v1/chain/utxos/byids\?id\=08781f222444deb6a672972b543b05bd27f927a2938e08789f744b5034dc24b914
[
  {
    "output_type": "Transaction",
    "commit":
 [8,120,31,34,36,68,222,182,166,114,151,43,84,59,5,189,39,249,39,162,147,142,8,120,159,116,75,80,52,220,36,185,20],
    "switch_commit_hash": null,
    "proof": null,
    "height": 4041,
    "lock_height": 0
  }
]
```

In contrast we serialize like this for the blocks api endpoint - 
```
{
    "header": {
        "hash": "20eeb00bb8ee8010fc9ade307270cf272d73400487a6c104e8d038aa981924ae",
        "version": 1,
        "height": 4042,
        "previous": "edcb4ebe14f3f779d4ccadd12834d9242fbff7d3b47c40bb31c86bea8d59fadb",
        "timestamp": "2017-12-28T15:44:46Z",
        "utxo_root": "c167a975305a5ef7928f6e144ee1a50a70b1c8160042f07ef4ab29640afd907f",
        "range_proof_root": "c70c745ae2ce81460fb6cab55436418a91fd45e4719eecbe23d2439864eceddd",
        "kernel_root": "d018328db62e99ca5cf70bfe40a14d36ba671a6b3cb2d9cb8c81ca4d76014a31",
        "nonce": 13328075940776718580,
        "difficulty": 544,
        "total_difficulty": 12933876
    },
    "inputs": [
        "08974bcf0bdcabdb56f10959e35cab6917b0419a7c951398ef67c68fb660f4765f",
        "08a94fd0bddec0d713f24ae02d82be0ef8f4ed7e8ab65d0096317adedb1a5966ba",
        "08781f222444deb6a672972b543b05bd27f927a2938e08789f744b5034dc24b914",
        "099d4178e99d37e92282e35ba6cd3c95b0c6cbdbf7fddf5469604051dcc4829e0a",
        "08a8c9f6dc3666cd486fa0011da40d94f2e6652d6a32e7121b46087d54b1fd5425",
        "089c63852e08c6797bcb5a7734c0f62ebca565a5c74d86d299e24dbfe54e900c3b"
    ],
    "outputs": [
        {
            "output_type": "Transaction",
            "commit": "08c4d5f74b74ab29db67bb6d660f015ee57f0f509eef8aa3932ac716ab6e34ad82",
            "switch_commit_hash": "1a182bdcfd345cd94be1c3141327d94d295c0de8",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "c1947aaa2613694af2056ce5930c4caf3ed38e2d29e68ae2ac0aad7e370c74a5"
        },
        {
            "output_type": "Transaction",
            "commit": "08bae25fb3e58624054d9560c61e40124f6e5a5aee0b0e407233cd11e4a150861d",
            "switch_commit_hash": "94bd27947af66b48492147deea237bcd8808688f",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "cf62fb4ef0cbd2ba86066262f7fffd360e6d2092ab3b9557fec5c297d31276db"
        },
        {
            "output_type": "Transaction",
            "commit": "09ed25980a656639ba837655bf4efb04a7f0fadcfccbeb26ec2fbd7d28affccab6",
            "switch_commit_hash": "d9b64c03b6c4a1fff90b2cd1c2aab26d7de021b9",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "daf848cf6ae851c9f864d3f60aaa049181db46f7e5ebb3b02eb03abf4b105054"
        },
        {
            "output_type": "Coinbase",
            "commit": "08d239587f18c3606417cc2a214746f6baaf6f95d1524592079668ae363c54056c",
            "switch_commit_hash": "0cab89d3e9f29aacd71bfccd3bb7339e8075a5b3",
            "height": 4042,
            "lock_height": 5042,
            "spent": true,
            "proof_hash": "abd76ccf8986b9d36c6a67947b79ea22611dfa5b56d1f7d6e09aae70f4d2cf49"
        },
        {
            "output_type": "Transaction",
            "commit": "08b1a6adfafc17e1aba3605fee776d1b84a3f57148fb20fe444eace49254855ac9",
            "switch_commit_hash": "7a45d4fffb59b210140c7e1abecbe0a70321224e",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "739ac80294932eb80086c6f2715f84597426b2627c6c5f7bbe53a2d806ec6800"
        },
        {
            "output_type": "Transaction",
            "commit": "09d2ccdb05280570337a2633b5edef71099210830fe4b00fa63b0f2955afbc9bda",
            "switch_commit_hash": "42d2c1e72398aeaeec81443c47f988f2e4c8baf3",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "40d1d0d31f20ae1c504816e5358f2267a0c7c1bfdeaa51e840e2736022d1c69c"
        },
        {
            "output_type": "Transaction",
            "commit": "0828bfafe8a8c3b6ca248f1c8307be6278f7b08b70d76c84cc032f579712bf6589",
            "switch_commit_hash": "fe676cd9691b88f868db923bf042f249767bd70a",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "736def670dd48b62e5cabb58be4c4260bb362db37928c0e6d25a8cd556d53d62"
        },
        {
            "output_type": "Transaction",
            "commit": "086118f9d2348985648cfe33112a1e800951dbff42c3d80c2868745745509c7bb9",
            "switch_commit_hash": "d06192a44c7d877c40422581ffd0a9d0ce436065",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "563bafca0fa8ad51598433d93d0265d07ded22a3a4968d52de2d3800dfc8876e"
        },
        {
            "output_type": "Transaction",
            "commit": "083fd38d4c95ef6199c5f7ca3ec49b6e28c7e9c2483fe717c20cb18f3089bbd6da",
            "switch_commit_hash": "2bbecf0138c1c00d360fb66aabe9a87e2bfd8883",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "0fe3f74775bedc1ce4205b219f2d2575bef73d80ccd34156f2594e42acc650e9"
        },
        {
            "output_type": "Transaction",
            "commit": "09da32ca0664083a31b4840cfa2bc4303e852f2b408983b5d8ba1da25b9c9da6e9",
            "switch_commit_hash": "9f85336ad71e2578e7366f3e954232a850368cd4",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "4dfa15cdfcf66171d544fa41f813813c3637023a1c383e67ed7d0fc509b5f04f"
        },
        {
            "output_type": "Transaction",
            "commit": "08bee227b967a2725f5eb4cb9c206d25b1ba4d1178fd0828bcf99565614df6cb23",
            "switch_commit_hash": "0495449875a8af172184ab95dc8408757bafdeb6",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "9dad2d7616b18297cf7197fea0483a7d21f34cefdec92ec5c9dc43d9c0e8722c"
        },
        {
            "output_type": "Transaction",
            "commit": "0813981bad1e849d5371ce5a6cb8178a28b9979064c314cfcd0303291f1f3cbb7f",
            "switch_commit_hash": "1703523ff45d1c8bbeec6b45c434ad9a1525ebe1",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "4b0a4f8bf0ebfbc8a3d8019e01deab428e99e1f694cbc7e99470fd3867d65d82"
        },
        {
            "output_type": "Transaction",
            "commit": "0872e351b72aa61682aef71e31e6e48692ca489cb8e76e66c1ad64ecb593b542ac",
            "switch_commit_hash": "1831727103bbfad740bf1c0ed3f84b419d93d0d5",
            "height": 4042,
            "lock_height": 0,
            "spent": true,
            "proof_hash": "ccbc8b1dcf456c0c61d252b644b23edd21365d7041f8253a4ff0bb190b7ffa07"
        }
    ],
    "kernels": [
        {
            "features": "",
            "fee": 8000000,
            "lock_height": 4041,
            "excess": "08539e24f76543f2b654a06973f6f11e9de99ce1ace4976f1029a6f39519a304c7",
            "excess_sig": "3045022100d36766d812fe9f754c345513af31dbf17783a2290a82030f6a111346fba67b91022074e2b21823ac3e416271f6b2ade9a395c2801e26c5f9da5f03b353340719ba9a"
        },
        {
            "features": "",
            "fee": 8000000,
            "lock_height": 4041,
            "excess": "097d38e16a2a6dce892867b58f6529d145383fd8d29ca8f16c900b8671fdd547e4",
            "excess_sig": "3045022100f5b2c7ef954bb3c0ac3404a1163a987d8f9d64e7b44f6b34053063e78c743bf1022072cc625db2209bbbb0dba8e906b34b9710a941f264c1f802268b5428b2f100f3"
        },
        {
            "features": "",
            "fee": 8000000,
            "lock_height": 4041,
            "excess": "08f86ed23aa9f28b8b9b66945894b216b95bb352d0f6012e7e41183bcae12555cf",
            "excess_sig": "3044022046e85075956d21608311f345a2f6a536c6240e30e52c603f26607810c3a4e9b602207481f0dd76186950a4fb0cf54ab12479a2b43673eaab65e7a8ccb50b5e4a8c78"
        },
        {
            "features": "COINBASE_KERNEL",
            "fee": 0,
            "lock_height": 0,
            "excess": "0965cd7909d9df2bef138eaf34d52d13c1258614dfc63e64ef247cee8629156b00",
            "excess_sig": "30450221009cbf768a91d298f6022311dc76589497f15d694c856b4978ff0a5cdd1218805f02207d4f6ba6ebaef7ba895f5247ae76a92ac7079747d50541b7d178b60671ab2e3a"
        },
        {
            "features": "",
            "fee": 8000000,
            "lock_height": 4040,
            "excess": "095a849eed110cb125be0ebd53eb39ac324a6d3de1667402e4122e391463e7bad0",
            "excess_sig": "304402204589372210a7a755e2325439e0a398ea9dc603defee81c3479aea92c7b1b8c3302203ec514e38a67ecd7760cf82bcb2392234980e68375205da49d8b2fed64ee1802"
        },
        {
            "features": "",
            "fee": 8000000,
            "lock_height": 4040,
            "excess": "0963aeaee87ef7505ece0ad11736c4593f5bc76e286913a11fbd1b94bfc194e306",
            "excess_sig": "3045022100e20092d26f8d760f40de80148d5ca25487b022f7743b963a483a64ed87ae2c70022030550bc387e42e804dc5649e1a42a83902bc7a624dfcfe6562d8049c9e5e11e9"
        },
        {
            "features": "",
            "fee": 8000000,
            "lock_height": 4041,
            "excess": "091ec6c2d7e1c3e0f4ce462f20537caf97ba7f41b4d0c1d50f4d6eb4ab59d66dc9",
            "excess_sig": "3044022023f2551266018226a4adc6ac8004b781529b2ccbf860b61c6678c68704a4459402200f9b02932ad0f994ef6b429aaaef940c8af46723fd0a447c053cfc979c3aff24"
        }
    ]
}
```

### PR DESCRIPTION
**Pull request by mslipper**

### CODE REVIEW

**Comment by antiochp** [comment]
We have the following representations of an `Output` in the api code - 
* Output
* OutputPrintable
* OutputSwitch

Seems like we may be able to consolidate these into just a single representation (this is likely what you are doing by using OutputPrintable and not Output).

As these are all http/json serialization (via the api) - it seems like everything should be "printable".
I don't think there is much need for non-printable versions of an output (whatever that actually means).

**Comment by mslipper** [comment]
I can give that a shot. Full disclosure, I'm still a Rust noob so it'll take a little time for me to refactor all the places that these output types are used and to test them thoroughly.

**Comment by ignopeverell** [comment]
Any interest in picking this up?

**Comment by antiochp** [comment]
I'm actually deep in the output serialization code due to us picking #215 back up.
We are getting rid of the "output_by_commit" index - so we have _very_ limited info available via the api when retrieving outputs by their commitments.
The _only_ way to get heights, output_types etc. now will be to go via the `byheight` api endpoint and look outputs up via their containing block.

So I'd recommend holding off on changing any of the output serialization until #215 is merged (assuming we don't find any blockers with this proposed approach).

**Comment by mslipper** [comment]
Ok great. I'll hold off then until #215 is merged.

**Comment by antiochp** [comment]
Ok - #215 ended up not getting merged. 
_But_ - we merged #615 instead.

So now we have the following "models" in api - 
* `OutputPrintable` (everything is basically a string here)
* `Utxo` (just a wrapper for a commitment, replaces the old `OutputSwitch` model)

A Utxo currently looks like this -
```
curl "127.0.0.1:13413/v1/chain/utxos/byids?id=0960dd49bc81aa33fe094d98126ad3f759eb0ff5ee36c75ae82acd2486253df18f"
[{"commit":[9,96,221,73,188,129,170,51,254,9,77,152,18,106,211,247,89,235,15,245,238,54,199,90,232,42,205,36,134,37,61,241,143]}]
```

And OutputPrintables look like this (nested in a BlockPrintable here) - 
```
curl "127.0.0.1:13413/v1/chain/utxos/byheight?startheight=1&endheight=1"
[
  {
    "header":{
      "hash":"7272a182aed9268920d4949929490511d2006864ee5bdd976d3a4595cf1c8f94",
      "height":1,
      "previous":"b83053b4b2e29376d6850de4fdaa60366300520bc6957cc69bec7089fcb1bfd8"
    },
    "outputs":[
      {
        "output_type":"Coinbase",
        "commit":"083eafae5d61a85ab07b12e1a51b3918d8e6de11fc6cde641d54af53608aa77b9f",
        "switch_commit_hash":"85daaf11011dc11e52af84ebe78e2f2d19cbdc76",
        "spent":false,
        "proof":null,
        "proof_hash":"ed6ba96009b86173bade6a9227ed60422916593fa32dd6d78b25b7a4eeef4946"
      }
    ]
  }
]
```

So - 
1) I think we should serialize the commit in the Utxo as a hex string (in the json).
2) I do think we are currently approaching the serialization problem the wrong way in the api code - OutputPrintable (and other api models) should maintain its internal state using Commitment, Hash etc. And we should serialize these to hex strings when necessary (rather than maintaining state internally in strings etc.)

**Comment by mslipper** [comment]
Agreed on your points. I can pick this back up and look into using Rust's `serde` library to automatically convert between hex internal representations of these objects.

**Comment by antiochp** [comment]
Sounds like a good plan.
We're doing similar stuff in the wallet code (to serialize outputs out to wallet.dat).

For example - 
https://github.com/mimblewimble/grin/blob/f1bbf5370edd2d5f9a6d233e976495ed2ba3be9c/wallet/src/types.rs#L246-L253

**Comment by mslipper** [comment]
Great, this is really helpful!

**Comment by mslipper** [comment]
@antiochp FYI will have this ready by EOD today.

File: api/src/types.rs [hunk 1]
```diff
@@ -429,3 +542,18 @@ pub struct PoolInfo {
 	/// Total size of pool + orphans
 	pub total_size: usize,
 }
+
+#[test]
+fn serialize_printable_commitment() {
+	let hex_output = "{\
+        \"output_type\":\"Coinbase\",\
+        \"commit\":\"083eafae5d61a85ab07b12e1a51b3918d8e6de11fc6cde641d54af53608aa77b9f\",\
+        \"switch_commit_hash\":\"85daaf11011dc11e52af84ebe78e2f2d19cbdc76000000000000000000000000\",\
```
Comment (mslipper) [inline api/src/types.rs]: @antiochp I observed that the hex values that were being outputted were only 160 bits long while the switch commit hash is supposed to be 256 bits long. In this implementation, the switch commit length is 256 bits as per `transaction.rs`.

File: api/src/types.rs [hunk 2]
```diff
@@ -153,29 +167,72 @@ pub enum OutputType {
 #[derive(Debug, Serialize, Deserialize, Clone)]
 pub struct Utxo {
 	/// The output commitment representing the amount
-	pub commit: pedersen::Commitment,
+	pub commit: PrintableCommitment,
```
Comment (mslipper) [inline api/src/types.rs]: There is already a serializer defined for the pedersen commitment, so I had to create a small newtype struct to wrap it. Let me know if there is a better way to do this.

Comment (antiochp) [reply api/src/types.rs]: There _may_ be a better way of doing this in serde but I don't know what it is...
This looks fine to me.

File: api/src/types.rs [hunk 3]
```diff
@@ -197,44 +254,143 @@ impl OutputPrintable {
...
 		};
 
 		OutputPrintable {
-			output_type: output_type,
-			commit: util::to_hex(output.commit.0.to_vec()),
-			switch_commit_hash: output.switch_commit_hash.to_hex(),
-			spent: spent,
-			proof: proof,
+			output_type,
+			commit: output.commit,
```
Comment (antiochp) [inline api/src/types.rs]: This feels a lot cleaner to me.

File: api/src/types.rs [hunk 4]
```diff
@@ -197,44 +254,143 @@ impl OutputPrintable {
...
+		state.serialize_field("commit", &util::to_hex(self.commit.0.to_vec()))?;
+		state.serialize_field("switch_commit_hash", &self.switch_commit_hash.to_hex())?;
+		state.serialize_field("spent", &self.spent)?;
+		state.serialize_field("proof", &self.proof)?;
+		state.serialize_field("proof_hash", &self.proof_hash)?;
+		state.end()
+	}
+}
+
+impl<'de> serde::de::Deserialize<'de> for OutputPrintable {
```
Comment (antiochp) [inline api/src/types.rs]: I guess this is the price we pay for telling serde to do the deserialization for us...

**Pull request merged**